_posts/2018-12-06.Rust-1.31.md

```diff
@@ -0,0 +1,338 @@
+    let mut x = 5;
+
+    let y = &x;
+
+    let z = &mut x;
+}
+```
+
+In older Rust, this is a compile-time error:
+
+```text
+error[E0502]: cannot borrow `x` as mutable because it is also borrowed as immutable
+ --> src/main.rs:5:18
+  |
+4 |     let y = &x;
+  |              - immutable borrow occurs here
+5 |     let z = &mut x;
+  |                  ^ mutable borrow occurs here
+6 | }
+  | - immutable borrow ends here
+```
+
+This is because lifetimes follow "lexical scope"; that is, the borrow from `y`
+is considered to be held until `y` goes out of scope at the end of main, even
+though we never use `y` again. This code is fine, but the borrow checker could
+not handle it.
+
+Today, this code will compile just fine.
+
+What if we did use `y`, like this for example:
+
+```rust
+fn main() {
+    let mut x = 5;
+    let y = &x;
+    let z = &mut x;
+    
+    println!("y: {}", y);
+}
+```
+
+Older Rust will give you this error:
+
+```text
+rror[E0502]: cannot borrow `x` as mutable because it is also borrowed as immutable
+ --> src/main.rs:5:18
+  |
+4 |     let y = &x;
+  |              - immutable borrow occurs here
+5 |     let z = &mut x;
+  |                  ^ mutable borrow occurs here
+...
+8 | }
+  | - immutable borrow ends here
+```
+
+With Rust 2018, this error changes for the better:
+
+```text
+error[E0502]: cannot borrow `x` as mutable because it is also borrowed as immutable
+ --> src/main.rs:5:13
+  |
+4 |     let y = &x;
+  |             -- immutable borrow occurs here
+5 |     let z = &mut x;
+  |             ^^^^^^ mutable borrow occurs here
+6 |     
+7 |     println!("y: {}", y);
+  |                       - borrow later used here
+```
+
+Instead of pointing to where `y` goes out of scope, it shows you where the
+conflicting borrow occurs. This makes these sorts of errors far easier to
+debug.
+
+In Rust 1.31, this feature is exclusive to Rust 2018. We hope to backport it
+to Rust 2015 at a later date.
+
+#### Module system changes
+
+The module system is often one of the hardest things for people new to Rust.
+Everyone has their own things that take time to master, of course, but
+there's a root cause for why it's so confusing to many: while there are
+simple and consistent rules defining the module system, their consequences
+can feel inconsistent, counterintuitive and mysterious.
+
+As such, the 2018 edition of Rust introduces a few new module system
+features, but they end up simplifying the module system, to make it more
+clear as to what is going on.
+
+Here's a brief summary:
+
+* `extern crate` is no longer needed in 99% of circumstances.
+* You can import macros with `use`, rather than a `#[macro_use]` attribute.
+* Absolute paths begin with a crate name, where the keyword `crate` refers to the current crate.
+* A `foo.rs` and `foo/` subdirectory may coexist; `mod.rs` is no longer needed when placing submodules in a subdirectory.
+
+These may seem like arbitrary new rules when put this way, but the mental
```

Comment:
This feels a bit underwhemling. I'm wondering if we can get across a bit more about why this is an improvement. For example, we could talk about how this eliminates the confusion around paths working differently at the top level module vs submodules.

Reply:
I was worried about the post being too long.... doing that adds a *lot* of length. I agree that it feels underwhelming though...